doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,159 @@
+
+For now, I would propose to implement only the soft limit. [Later we can consider implementing also the hard limit.](https://github.com/MaterializeInc/materialize/issues/18832) Note that implementing the hard limit will be easy to do after having implemented the soft limit.
+
+## Syntax
+
+An example for the proposed syntax:
+```
+WITH MUTUALLY RECURSIVE (ITERATION LIMIT 100)
+  cnt (i int) AS (
+    SELECT 1 AS i
+    UNION
+    SELECT i+1 FROM cnt)
+SELECT * FROM cnt;
+```
+
+Our WMR implementation supports having multiple WMR blocks inside a single query. An example is computing [strongly connected components](https://github.com/MaterializeInc/materialize/blob/9037f58c398c029924bbb59c68a3e6278a6a6f25/test/sqllogictest/with_mutually_recursive.slt#L222-L274), which can be implemented by nested WMRs. In case of multiple WMR blocks, the above syntax allows for setting different limits for each WMR block.
+
+Having separate limits for separate WMR blocks in one query is especially important for the **compositionality** of views. Let's say that Alice writes a view that has just a single WMR. Then Bob writes another query that uses Alice's view, and has a WMR itself. The view is inlined, so Bob's query now has two WMRs. But Bob shouldn't need to care about the internals of Alice's view, and might not even know that the view has a WMR, so he thinks his WMR is the only one. And then he sets a limit for the entire query, which will unexpectedly mess up Alice's WMR that is inlined from inside the view.
+
+The syntax follows [our SQL Design Principles](https://www.notion.so/materialize/SQL-design-principles-2e2069797f7c46b8b4c7dd10043afdcd):
+- The options block is parenthesized.
+- There is a space between ITERATION and LIMIT.
+- We accept an optional `=` before the number.
+- We use the `generate_extracted_config!` macro to process the options.
+
+The exact keywords are not final yet, and also depends on what the exact keyword will be for WITH MUTUALLY RECURSIVE.
+
+## Rendering
+
+The rendering part of the design is simple and [uncontroversial](https://materializeinc.slack.com/archives/C041KPFCR98/p1679655472232159?thread_ts=1679653114.252349&cid=C041KPFCR98). We will use `branch_when` ([from Timely](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/branch.rs#L94)) to access the pointstamps containing the iteration numbers (after [the consolidation of the result of an iteration](https://github.com/MaterializeInc/materialize/blob/26f23fd7271cf92c7c8b0dc4ba4711999b45999a/src/compute/src/render/mod.rs#L643)). This will give us two streams: rows that are within the limit, and rows that are outside. The latter one we can either throw away or route into the error stream, depending on whether we want to just gracefully stop at the limit with the current state as the final result (soft limit), or error out when reaching the limit (hard limit).
```

Comment:
nit: replace the "from Timely" URL with a permalink:

https://github.com/TimelyDataflow/timely-dataflow/blob/432ef57fae761f1e4773833b0474b41f1efe7e7c/timely/src/dataflow/operators/branch.rs#L94

Reply:
fixed